integration-tests/capacity/replenishment.test.ts

```diff
@@ -0,0 +1,168 @@
+    const [setEpochLengthEvent] = await setEpochLengthOp.sudoSignAndSend();
+    const epoch_was_set = setEpochLengthEvent && ExtrinsicHelper.api.events.capacity.EpochLengthUpdated.is(setEpochLengthEvent);
+    assert(epoch_was_set, "failed to set epoch");
+    await createGraphChangeSchema();
+  });
+
+  describe("Capacity is replenished", function(){
+    it("after new epoch", async function() {
+      const totalCapacity = 3n*1500n*1000n;
+      const [stakeKeys, stakeProviderId] = await createAndStakeProvider("ReplFirst", totalCapacity);
+      const payload = JSON.stringify({ changeType: 1,  fromId: 1, objectId: 2 })
+      // const call = ExtrinsicHelper.api.tx.messages.addOnchainMessage(null, schemaId, payload);
+      const call = ExtrinsicHelper.addOnChainMessage(stakeKeys, schemaId, payload);
+
+      // confirm that we start with a full tank
+      let remainingCapacity = (await getRemainingCapacity(stakeProviderId)).toBigInt();
+      assert.equal(totalCapacity, remainingCapacity);
+
+      await call.payWithCapacity(-1);
+      remainingCapacity = (await getRemainingCapacity(stakeProviderId)).toBigInt();
+      assert(totalCapacity > remainingCapacity);
+
+      // one more txn to deplete capacity more so this current remaining is different from when
+      // we submitted the first message.
+      await call.payWithCapacity(-1);
+      const newEpochBlock = await getNextEpochBlock();
+      await ExtrinsicHelper.run_to_block(newEpochBlock);
+
+      // this should cause capacity to be refilled and then deducted by the cost of one message.
+      await call.payWithCapacity(-1);
+      let newRemainingCapacity = (await getRemainingCapacity(stakeProviderId)).toBigInt();
+
+      // this should be the same as after sending the first message, since it is the first message after
+      // the epoch.
+      assert.equal(remainingCapacity, newRemainingCapacity);
+    });
+  });
+
+  describe("Capacity is not replenished", function() {
+    it("if out of capacity and last_replenished_at is <= current epoch", async function() {
+      let [stakeKeys, stakeProviderId] = await createAndStakeProvider("NoSend", 3n*1000n*1000n);
+      let payload = JSON.stringify({ changeType: 1,  fromId: 1, objectId: 2 })
+      let call = ExtrinsicHelper.addOnChainMessage(stakeKeys, schemaId, payload);
+
+      // run until we can't afford to send another message.
+      await call.payWithCapacity(-1);
+
+      let nextEpochBlock = await getNextEpochBlock();
+
+      // mine until just before epoch rolls over
+      await ExtrinsicHelper.run_to_block(nextEpochBlock - 1);
+
+      // TODO: it's weird that we can run one more capacity txn before running out.
+      await call.payWithCapacity(-1)
+
+      await assert.rejects(
+        call.payWithCapacity(-1), {name: "RpcError", message: "1010: Invalid Transaction: Inability to pay some fees , e.g. account balance too low"});
```

Comment:
nit: comparing to lengthy error message strings verbatim is problematic, if source errors are updated for wording, spelling, etc. I find it better to use a regex for the key part of the message.